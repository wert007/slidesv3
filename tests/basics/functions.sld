func sum(values: int[]) {
    let result = 0;
    for value in values {
        result = result + value;
    }
    print(result);
}

func count_up(n: int) {
    if n > 0 {
        count_up(n - 1);
    }
    print(n);
}

func add(a: int, b: int) -> int {
    if a == 0 {
        return b;
    } else if b == 0 {
        return a;
    } else {
        return a + b;
    }
}

func sub(a: int, b: int) -> int {
    return a - b;
}

func mul(a: int, b: int) -> int {
    return a * b;
}

func div(a: int, b: int) -> int {
    return a / b;
}

func fac(n: int) -> int {
    if n <= 1 {
        return 1;
    } else {
        return n * fac(n - 1);
    }
}

func main() {
    let basic_arith = [
        add,
        sub,
        mul,
        div,
        // Throws rightfully diagnostic
        // fac,
    ];

    for index, arithmetic_function in basic_arith {
        let result = arithmetic_function(5, 3);
        print('fn(5, 3) = ' + result);
        print('fn(fn(5, 3), ' + index + ') = ' + arithmetic_function(result, index));
    }
}