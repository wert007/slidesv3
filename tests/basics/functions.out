fn(5, 3) = 8
fn(fn(5, 3), 0) = 8
fn(5, 3) = 2
fn(fn(5, 3), 1) = 1
fn(5, 3) = 15
fn(fn(5, 3), 2) = 30
fn(5, 3) = 1
fn(fn(5, 3), 3) = 0
= 0
